Debug CreateBlockFromBoundary functions

Feuerbowle opened the pull request:
The CreateBlockFromBoundary and CreateBlockFromBoundaryWithLimits functions now checking if the boundary exists before creating the block.

The CreateBlockFromBoundaryWithLimits now doesn't reset the block (isn't needed because we have the ZMin and ZMax)

Added an function to Export the Block as .dmt or .stl file

Feuerbowle committed:
Debug CreateBlockFromBoundary functions

The CreateBlockFromBoundary and CreateBlockFromBoundaryWithLimits functions now checking if the boundary exists before creating the block.

The CreateBlockFromBoundaryWithLimits now doesn't reset the block (isn't needed because we have the ZMin and ZMax)

Added an function to Export the Block as .dmt or .stl file

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMProjectTests.cs:
```diff
@@ -147,6 +147,24 @@ public void CreateBlockFromBoundaryTest()
+            _powerMILL.ActiveProject.Refresh();
+            PMBoundary boundary = _powerMILL.ActiveProject.Boundaries.ActiveItem;
+            var boundingBox = _powerMILL.ActiveProject.CreateBlockFromBoundaryWithLimits(boundary, 0, 100);
+            FileSystem.File file = new FileSystem.File(TestFiles.DIRECTORY + "\\ExportBlock.dmt");
```
I would suggest using a file in the temporary directory instead.  The File class has a method on it to give you a name of a temporary file.

Feuerbowle replied:
learned something new every day.

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMProject.cs:
```diff
@@ -429,13 +429,16 @@ public void CreateBlock(string modelName, double blockExpansion)
-            _powerMILL.DoCommand("EDIT BLOCK RESET");
-            _powerMILL.DoCommand("BLOCK ACCEPT");
+        {
+            if (boundary != null)
```
This feels weird...  If the parameter is null then you should throw a [ArgumentNullException](https://docs.microsoft.com/en-us/dotnet/api/system.argumentnullexception?view=netframework-4.8).

If the real issue is that the PMBoundary doesn't exist in PowerMill then use the Exists function to check for it.

Feuerbowle replied:
We have to check both. Because if the user call the function in this way

`powermill.ActiveProject.CreateBlockFromBoundary(powermill.ActiveProject.Boundaries.GetByName("Test"))`

the boundary is null if the boundary doesn't exists

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMProject.cs:
```diff
@@ -447,19 +450,33 @@ public BoundingBox CreateBlockFromBoundary(PMBoundary boundary)
+        /// <param name="file">The file where to save the block (.dmt and .stl only)</param>        
+        public void ExportBlock(Autodesk.FileSystem.File file)
+        {                      
+            if (file.Extension.ToLower() == "dmt" || file.Extension.ToLower() == "stl")
```
If the extension is something else then should we throw an exception?

Feuerbowle committed:
Minor changes

- Create a temporary file when testing the ExportBlock function
- Throw an exception if the file format is not supportet at the ExportBlock function

- Check if boundary is null and throw an exception if so at the CreateBlockFromBoundary and CreateBlockFromBoundaryWithLimits function

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMProject.cs:
```diff
@@ -430,7 +430,10 @@ public void CreateBlock(string modelName, double blockExpansion)
+            if (boundary == null)
+                throw new ArgumentNullException("boundary", "Boundary not found");
+
+            if (boundary.Exists)
```
At this point, I think you should combine the two.  Have:
`if (boundary == null || !boundary.Exists)`

lukeedw commented:
Made, one comment, other than that I think it looks good to go.

Feuerbowle committed:
Update PMProject.cs

combined the two ifs
